src/foam/nanos/ruler/RulerDAO.js

```diff
@@ -123,14 +172,93 @@
         }
       ],
       javaCode: `
-
       for ( Object key : sink.getGroupKeys() ) {
         List<Rule> group = ((ArraySink) sink.getGroups().get(key)).getArray();
         if ( ! group.isEmpty() ) {
+          Collections.sort(group, new foam.mlang.order.Desc(Rule.PRIORITY));
           new RuleEngine(x, this).execute(group, obj, oldObj);
         }
       }
       `
+    },
+    {
+      name: 'updateRules',
```

Comment:
If we fetch relevant rules each time when applying this won't be needed.

Reply:
Too slow.

Reply:
Correct me if I'm wrong. The slow part isn't on fetching rules from DAO but on executing the rules themselves. The performance gain wouldn't be noticeable considering JDAO is already optimized.

Reply:
No, executing the rules is probably fast because it just applies a predicate, which doesn't create any objects or generate any garbage. It's O(1). Fetching objects from a DAO is O(lnn) plus generates objects when they are cloned and then needs to garbage collect. Plus that wouldn't help us because we have stateful rules so need to use the same one every time.